src/common/helper/ShapeHelper.cpp

```diff
@@ -127,23 +129,59 @@ QPointF ShapeHelper::rectPointAtIndex(const QRectF &rect, int index)
 	}
 }
 
-QRectF ShapeHelper::setRectPointAtIndex(const QRectF &rect, int index, const QPointF &pos)
+QRectF ShapeHelper::setRectPointAtIndex(const QRectF &rect, int index, const QPointF &pos, bool keepAspectRatio)
 {
 	auto updatedRect = rect;
 	if (index == 0) {
-		updatedRect.setTopLeft(pos);
+		if (!keepAspectRatio) {
+			updatedRect.setTopLeft(pos);
+		} else {
+			const auto xDif = updatedRect.topLeft().x() - pos.x();
+			const auto yDif = updatedRect.topLeft().y() - pos.y();
+			const auto min = qMin(xDif, yDif);
+			const auto newX = updatedRect.topLeft().x() - min;
+			const auto newY = updatedRect.topLeft().y() - min;
+			updatedRect.setTopLeft(QPointF(newX, newY));
```

Comment:
Don't we have some similar logic in the addPoint for Rect? I believe we can already keep aspect ratio when drawing the initial rect, could that have somehow been used for this feature? If not, could extract this logic in a separate class? I imagine we could have something like a RectShapeUpdater that gets just a rect and a bool for the aspectRatio as parameter and either updates that rect or returns a new updated rect  (probably rather the later). You would have here only a call to this class and this class could be simply covered by unit tests.